commit 5548770b5f8c31502428e663fbcecc2fb372900c
Author: Seth Forshee <seth.forshee@canonical.com>
Date:   Thu Oct 2 15:34:45 2014 -0500

    UBUNTU: SAUCE: (namespace) fuse: Restrict allow_other to the superblock's namespace or a descendant
    
    Unprivileged users are normally restricted from mounting with the
    allow_other option by system policy, but this could be bypassed
    for a mount done with user namespace root permissions. In such
    cases allow_other should not allow users outside the userns
    to access the mount as doing so would give the unprivileged user
    the ability to manipulate processes it would otherwise be unable
    to manipulate. Restrict allow_other to apply to users in the same
    userns used at mount or a descendant of that namespace. Also
    export current_in_userns() for use by fuse when built as a
    module.
    
    Acked-by: Serge Hallyn <serge.hallyn@canonical.com>
    Acked-by: Miklos Szeredi <mszeredi@redhat.com>
    Signed-off-by: Seth Forshee <seth.forshee@canonical.com>
    Signed-off-by: Tim Gardner <tim.gardner@canonical.com>

diff --git a/fs/fuse/dir.c b/fs/fuse/dir.c
index 2a74a07..242d975 100644
--- a/fs/fuse/dir.c
+++ b/fs/fuse/dir.c
@@ -1020,7 +1020,7 @@ int fuse_allow_current_process(struct fuse_conn *fc)
 	const struct cred *cred;
 
 	if (fc->flags & FUSE_ALLOW_OTHER)
-		return 1;
+		return current_in_userns(fc->user_ns);
 
 	cred = current_cred();
 	if (uid_eq(cred->euid, fc->user_id) &&
diff --git a/kernel/user_namespace.c b/kernel/user_namespace.c
index 2ab4c47..7467a1e 100644
--- a/kernel/user_namespace.c
+++ b/kernel/user_namespace.c
@@ -957,6 +957,7 @@ bool current_in_userns(const struct user_namespace *target_ns)
 	}
 	return false;
 }
+EXPORT_SYMBOL(current_in_userns);
 
 static inline struct user_namespace *to_user_ns(struct ns_common *ns)
 {
